monai/data/meta_obj.py

```diff
@@ -0,0 +1,207 @@
+        `MetaObj`.
+
+        Args:
+            args: Sequence of inputs to be flattened.
+        Returns:
+            list of nested `MetaObj` from input.
+        """
+        out = []
+        for a in args:
+            if isinstance(a, (list, tuple)):
+                out += MetaObj.flatten_meta_objs(a)
+            elif isinstance(a, MetaObj):
+                out.append(a)
+        return out
+
+    def _copy_attr(self, attribute: str, input_objs: list[MetaObj], default_fn: Callable, deep_copy: bool) -> None:
+        """
+        Copy an attribute from the first in a list of `MetaObj`. In the case of
+        `torch.add(a, b)`, both `a` and `b` could be `MetaObj` or something else, so
+        check them all. Copy the first to `self`.
+
+        We also perform a deep copy of the data if desired.
+
+        Args:
+            attribute: string corresponding to attribute to be copied (e.g., `meta`).
+            input_objs: List of `MetaObj`. We'll copy the attribute from the first one
+                that contains that particular attribute.
+            default_fn: If none of `input_objs` have the attribute that we're
+                interested in, then use this default function (e.g., `lambda: {}`.)
+            deep_copy: Should the attribute be deep copied? See `_copy_meta`.
+
+        Returns:
+            Returns `None`, but `self` should be updated to have the copied attribute.
+        """
+        attributes = [getattr(i, attribute) for i in input_objs]
+        if len(attributes) > 0:
+            val = attributes[0]
+            if deep_copy:
+                val = deepcopy(val)
+            setattr(self, attribute, val)
+        else:
+            setattr(self, attribute, default_fn())
+
+    def _copy_meta(self, input_objs: list[MetaObj]) -> None:
+        """
+        Copy metadata from a list of `MetaObj`. For a given attribute, we copy the
+        adjunct data from the first element in the list containing that attribute.
+
+        If there has been a change in `id` (e.g., `a=b+c`), then deepcopy. Else (e.g.,
+        `a+=1`), then don't.
+
+        Args:
+            input_objs: list of `MetaObj` to copy data from.
+
+        """
+        id_in = id(input_objs[0]) if len(input_objs) > 0 else None
+        deep_copy = id(self) != id_in
+        self._copy_attr("meta", input_objs, self.get_default_meta, deep_copy)
+
+    def get_default_meta(self) -> dict:
+        """Get the default meta.
+
+        Returns:
+            default metadata.
+        """
+        return {}
+
+    def __repr__(self) -> str:
+        """String representation of class."""
+        out: str = super().__repr__()
+
+        out += "\nMetaData\n"
+        if self.meta is not None:
+            out += "".join(f"\t{k}: {v}\n" for k, v in self.meta.items())
+        else:
+            out += "None"
+
+        return out
+
+    @property
+    def meta(self) -> dict:
+        """Get the meta."""
+        return self._meta
```

Comment:
Hi @rijobro ,

Thanks for the good idea to construct `meta` property.
If `self._meta` is not initialized yet, I suggest to give a clear error message or return `get_default_meta()`?
What do you think?
Current error is hard to understand:
```py
class Test:
    meta: dict

    @property
    def meta(self) -> dict:
        """Get the meta."""
        return self._meta

    @meta.setter
    def meta(self, d: dict) -> None:
        """Set the meta."""
        self._meta = d


a = Test()
print(a.meta)
```
```
Traceback (most recent call last):
  File "/workspace/PythonTests/main.py", line 16, in <module>
    print(a.meta)
  File "/workspace/PythonTests/main.py", line 7, in meta
    return self._meta
AttributeError: 'Test' object has no attribute '_meta'
```

Thanks.